src/System.Threading.Thread/src/System/Threading/Thread.cs

```diff
@@ -0,0 +1,420 @@
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref sbyte address, sbyte value) => Volatile.Write(ref address, value);
+        public static void VolatileWrite(ref float address, float value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref ushort address, ushort value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref uint address, uint value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref ulong address, ulong value) => Volatile.Write(ref address, value);
+        [CLSCompliant(false)]
+        public static void VolatileWrite(ref UIntPtr address, UIntPtr value) => Volatile.Write(ref address, value);
+
+        /// <summary>
+        /// Manages functionality required to support members of <see cref="Thread"/> dealing with thread-local data
+        /// </summary>
+        private static class LocalDataStore
+        {
+            private static Dictionary<string, LocalDataStoreSlot> s_nameToSlotMap;
+
+            public static LocalDataStoreSlot AllocateSlot()
+            {
+                return new LocalDataStoreSlot(new ThreadLocal<object>());
+            }
+
+            public static Dictionary<string, LocalDataStoreSlot> EnsureNameToSlotMap()
+            {
+                var nameToSlotMap = s_nameToSlotMap;
+                if (nameToSlotMap != null)
+                {
+                    return nameToSlotMap;
+                }
+
+                nameToSlotMap = new Dictionary<string, LocalDataStoreSlot>();
+                return Interlocked.CompareExchange(ref s_nameToSlotMap, nameToSlotMap, null) ?? nameToSlotMap;
+            }
+
+            public static LocalDataStoreSlot AllocateNamedSlot(string name)
+            {
+                var slot = AllocateSlot();
+                var nameToSlotMap = EnsureNameToSlotMap();
+                lock (nameToSlotMap)
+                {
+                    nameToSlotMap.Add(name, slot);
+                }
+                return slot;
+            }
+
+            public static LocalDataStoreSlot GetNamedSlot(string name)
+            {
+                var nameToSlotMap = EnsureNameToSlotMap();
+                lock (nameToSlotMap)
+                {
+                    LocalDataStoreSlot slot;
+                    if (!nameToSlotMap.TryGetValue(name, out slot))
+                    {
+                        slot = AllocateSlot();
+                        nameToSlotMap[name] = slot;
+                    }
+                    return slot;
+                }
+            }
+
+            public static void FreeNamedSlot(string name)
+            {
+                var nameToSlotMap = EnsureNameToSlotMap();
+                lock (nameToSlotMap)
+                {
+                    nameToSlotMap.Remove(name);
```

Comment:
Looks like LocalDataStoreSlot wraps a ```ThreadLocal<T>```, and ```ThreadLocal<T>``` is IDisposable.  Can we dispose of it here, or is that risky for some reason?